nvtabular/inference/triton/__init__.py

```diff
@@ -0,0 +1,96 @@
+import os
+import subprocess
+from shutil import copyfile
+
+import cudf
+import tritonclient.http as httpclient
+from google.protobuf import text_format
+from tritonclient.utils import np_to_triton_dtype
+
+# read in the triton ModelConfig proto object - generating it if it doesn't exist
+try:
+    import nvtabular.inference.triton.model_config_pb2 as model_config
+except ImportError:
+    pwd = os.path.dirname(__file__)
+    try:
+        subprocess.check_output(
+            ["protoc", f"--python_out={pwd}", f"--proto_path={pwd}", "model_config.proto"]
+        )
+    except Exception as e:
+        raise ImportError("Failed to compile model_config.proto - is protobuf installed?") from e
+    import nvtabular.inference.triton.model_config_pb2 as model_config
+
+
+def generate_triton_model(workflow, name, output_path, version=1):
+    """ converts a workflow to a triton mode """
+    workflow.save(os.path.join(output_path, str(version), "workflow"))
+    _generate_model_config(workflow, name, output_path)
+    copyfile(
+        os.path.join(os.path.dirname(__file__), "model.py"),
+        os.path.join(output_path, str(version), "model.py"),
+    )
+
+
+def convert_df_to_triton_input(column_names, batch, input_class=httpclient.InferInput):
+    columns = [(col, batch[col]) for col in column_names]
+    inputs = [input_class(name, col.shape, np_to_triton_dtype(col.dtype)) for name, col in columns]
+    for i, (name, col) in enumerate(columns):
+        inputs[i].set_data_from_numpy(col.values_host)
+    return inputs
+
+
+def convert_triton_output_to_df(columns, response):
+    return cudf.DataFrame({col: response.as_numpy(col) for col in columns})
+
+
+def _generate_model_config(workflow, name, output_path):
```

Comment:
Did you update this function in this PR? If not, could you please check again?

This function in the "triton_python_poc" branch was generating a wrong formatted config file. Don't know how it works with Python backend but it gave me error when I tried with C++ backend.

This is what I am getting for Rossmann;
```
name: "rossmann"
backend: "python"
input {
  name: "Store"
  data_type: TYPE_INT64
  dims: -1
}
input {
  name: "DayOfWeek"
  data_type: TYPE_INT64
  dims: -1
}
input {
  name: "Year"
  data_type: TYPE_INT64
  dims: -1
}
...
...
output {
  name: "Store"
  data_type: TYPE_INT64
  dims: -1
}
output {
  name: "DayOfWeek"
  data_type: TYPE_INT64
  dims: -1
}
...
...
```
But, the format should be something like;

```
name: "rossmann"
backend: "python"
input [
{
  name: "Store"
  data_type: TYPE_INT64
  dims: -1
},
{
  name: "DayOfWeek"
  data_type: TYPE_INT64
  dims: -1
},
...
...
output [
{
  name: "Store"
  data_type: TYPE_INT64
  dims: -1
},
...
]
```

Reply:
This definitely works for the python backend -  whats the error you were seeing for the c++ backend? What version of protobuf are you using to load this config file?

Reply:
> The error is coming from the Triton itself. It’s saying the config file’s format is not correct. When you look at the config file examples, you can see that they are different that what is generated here. I don’t know why it works with python backend though.

I've built your C++ backend and changed the backend of one these model files - and triton loads up all the inputs/outputs from the model config file I write out here and dumps them out to the console.   However, I can't run your backend fully - its complaining about missing binaries etc. 

Can you:

 * Post the error text you're seeing
 * Post a dockerfile to build/run the c++ backend? I'm guessing this is just an environment difference, potentially with me running a newer version of protobuf

Fwiw I've included the triton server log here:

<details><summary>Triton server log</summary>
<p>

```
I0202 19:39:19.839251 1083 metrics.cc:221] Collecting metrics for GPU 0: GeForce GTX 1080 Ti
I0202 19:39:20.060508 1083 libtorch.cc:945] TRITONBACKEND_Initialize: pytorch
I0202 19:39:20.060536 1083 libtorch.cc:955] Triton TRITONBACKEND API version: 1.0
I0202 19:39:20.060540 1083 libtorch.cc:961] 'pytorch' TRITONBACKEND API version: 1.0
2021-02-02 19:39:20.204282: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcudart.so.11.0
I0202 19:39:20.242546 1083 tensorflow.cc:1877] TRITONBACKEND_Initialize: tensorflow
I0202 19:39:20.242595 1083 tensorflow.cc:1887] Triton TRITONBACKEND API version: 1.0
I0202 19:39:20.242610 1083 tensorflow.cc:1893] 'tensorflow' TRITONBACKEND API version: 1.0
I0202 19:39:20.242623 1083 tensorflow.cc:1917] backend configuration:
{}
I0202 19:39:20.244554 1083 onnxruntime.cc:1715] TRITONBACKEND_Initialize: onnxruntime
I0202 19:39:20.244569 1083 onnxruntime.cc:1725] Triton TRITONBACKEND API version: 1.0
I0202 19:39:20.244573 1083 onnxruntime.cc:1731] 'onnxruntime' TRITONBACKEND API version: 1.0
I0202 19:39:20.386073 1083 pinned_memory_manager.cc:199] Pinned memory pool is created at '0x7f329a000000' with size 268435456
I0202 19:39:20.386385 1083 cuda_memory_manager.cc:103] CUDA memory pool is created on device 0 with size 67108864
I0202 19:39:20.388399 1083 model_repository_manager.cc:787] loading: rossmann_categorify:1
I0202 19:39:20.488578 1083 model_repository_manager.cc:787] loading: rossmann_hashbucket:1
I0202 19:39:20.492359 1083 backend.cc:335] TRITONBACKEND_Initialize: nvtabular
I0202 19:39:20.492407 1083 backend.cc:348] Triton TRITONBACKEND API version: 1.0
I0202 19:39:20.492423 1083 backend.cc:354] 'nvtabular' TRITONBACKEND API version: 1.0
I0202 19:39:20.492439 1083 backend.cc:376] backend configuration:
{}
I0202 19:39:20.492936 1083 backend.cc:425] TRITONBACKEND_ModelInitialize: rossmann_categorify (version 1)
I0202 19:39:20.492958 1083 backend.cc:438] Repository location: /models/rossmann_categorify
I0202 19:39:20.492974 1083 backend.cc:451] backend state is 'backend state'
E0202 19:39:20.493220 1083 backend.cc:118] libpython3.8.so: cannot open shared object file: No such file or directory
I0202 19:39:20.499873 1083 backend.cc:217] model configuration:
{
    "name": "rossmann_categorify",
    "platform": "",
    "backend": "nvtabular",
    "version_policy": {
        "latest": {
            "num_versions": 1
        }
    },
    "max_batch_size": 0,
    "input": [
        {
            "name": "Store",
            "data_type": "TYPE_INT64",
            "format": "FORMAT_NONE",
            "dims": [
                -1
            ],
            "is_shape_tensor": false,
            "allow_ragged_batch": false
        },
        {
            "name": "DayOfWeek",
            "data_type": "TYPE_INT64",
            "format": "FORMAT_NONE",
            "dims": [
                -1
            ],
            "is_shape_tensor": false,
            "allow_ragged_batch": false
        },
        {
            "name": "Year",
            "data_type": "TYPE_INT64",
            "format": "FORMAT_NONE",
            "dims": [
                -1
            ],
            "is_shape_tensor": false,
            "allow_ragged_batch": false
        },
        {
            "name": "Month",
            "data_type": "TYPE_INT64",
            "format": "FORMAT_NONE",
            "dims": [
                -1
            ],
            "is_shape_tensor": false,
            "allow_ragged_batch": false
        },
        {
            "name": "Day",
            "data_type": "TYPE_INT64",
            "format": "FORMAT_NONE",
            "dims": [
                -1
            ],
            "is_shape_tensor": false,
            "allow_ragged_batch": false
        },
        {
            "name": "StateHoliday",
            "data_type": "TYPE_BOOL",
            "format": "FORMAT_NONE",
            "dims": [
                -1
            ],
            "is_shape_tensor": false,
            "allow_ragged_batch": false
        },
        {
            "name": "CompetitionMonthsOpen",
            "data_type": "TYPE_INT64",
            "format": "FORMAT_NONE",
            "dims": [
                -1
            ],
            "is_shape_tensor": false,
            "allow_ragged_batch": false
        },
        {
            "name": "Promo2Weeks",
            "data_type": "TYPE_INT64",
            "format": "FORMAT_NONE",
            "dims": [
                -1
            ],
            "is_shape_tensor": false,
            "allow_ragged_batch": false
        },
        {
            "name": "StoreType",
            "data_type": "TYPE_STRING",
            "format": "FORMAT_NONE",
            "dims": [
                -1
            ],
            "is_shape_tensor": false,
            "allow_ragged_batch": false
        },
        {
            "name": "Assortment",
            "data_type": "TYPE_STRING",
            "format": "FORMAT_NONE",
            "dims": [
                -1
            ],
            "is_shape_tensor": false,
            "allow_ragged_batch": false
        },
        {
            "name": "PromoInterval",
            "data_type": "TYPE_STRING",
            "format": "FORMAT_NONE",
            "dims": [
                -1
            ],
            "is_shape_tensor": false,
            "allow_ragged_batch": false
        },
        {
            "name": "CompetitionOpenSinceYear",
            "data_type": "TYPE_INT64",
            "format": "FORMAT_NONE",
            "dims": [
                -1
            ],
            "is_shape_tensor": false,
            "allow_ragged_batch": false
        },
        {
            "name": "Promo2SinceYear",
            "data_type": "TYPE_INT64",
            "format": "FORMAT_NONE",
            "dims": [
                -1
            ],
            "is_shape_tensor": false,
            "allow_ragged_batch": false
        },
        {
            "name": "State",
            "data_type": "TYPE_STRING",
            "format": "FORMAT_NONE",
            "dims": [
                -1
            ],
            "is_shape_tensor": false,
            "allow_ragged_batch": false
        },
        {
            "name": "Week",
            "data_type": "TYPE_INT64",
            "format": "FORMAT_NONE",
            "dims": [
                -1
            ],
            "is_shape_tensor": false,
            "allow_ragged_batch": false
        },
        {
            "name": "Events",
            "data_type": "TYPE_STRING",
            "format": "FORMAT_NONE",
            "dims": [
                -1
            ],
            "is_shape_tensor": false,
            "allow_ragged_batch": false
        },
        {
            "name": "Promo_fw",
            "data_type": "TYPE_INT64",
            "format": "FORMAT_NONE",
            "dims": [
                -1
            ],
            "is_shape_tensor": false,
            "allow_ragged_batch": false
        },
        {
            "name": "Promo_bw",
            "data_type": "TYPE_INT64",
            "format": "FORMAT_NONE",
            "dims": [
                -1
            ],
            "is_shape_tensor": false,
            "allow_ragged_batch": false
        },
        {
            "name": "StateHoliday_fw",
            "data_type": "TYPE_INT64",
            "format": "FORMAT_NONE",
            "dims": [
                -1
            ],
            "is_shape_tensor": false,
            "allow_ragged_batch": false
        },
        {
            "name": "StateHoliday_bw",
            "data_type": "TYPE_INT64",
            "format": "FORMAT_NONE",
            "dims": [
                -1
            ],
            "is_shape_tensor": false,
            "allow_ragged_batch": false
        },
        {
            "name": "SchoolHoliday_fw",
            "data_type": "TYPE_INT64",
            "format": "FORMAT_NONE",
            "dims": [
                -1
            ],
            "is_shape_tensor": false,
            "allow_ragged_batch": false
        },
        {
            "name": "SchoolHoliday_bw",
            "data_type": "TYPE_INT64",
            "format": "FORMAT_NONE",
            "dims": [
                -1
            ],
            "is_shape_tensor": false,
            "allow_ragged_batch": false
        },
        {
            "name": "CompetitionDistance",
            "data_type": "TYPE_FP64",
            "format": "FORMAT_NONE",
            "dims": [
                -1
            ],
            "is_shape_tensor": false,
            "allow_ragged_batch": false
        },
        {
            "name": "Max_TemperatureC",
            "data_type": "TYPE_INT64",
            "format": "FORMAT_NONE",
            "dims": [
                -1
            ],
            "is_shape_tensor": false,
            "allow_ragged_batch": false
        },
        {
            "name": "Mean_TemperatureC",
            "data_type": "TYPE_INT64",
            "format": "FORMAT_NONE",
            "dims": [
                -1
            ],
            "is_shape_tensor": false,
            "allow_ragged_batch": false
        },
        {
            "name": "Min_TemperatureC",
            "data_type": "TYPE_INT64",
            "format": "FORMAT_NONE",
            "dims": [
                -1
            ],
            "is_shape_tensor": false,
            "allow_ragged_batch": false
        },
        {
            "name": "Max_Humidity",
            "data_type": "TYPE_INT64",
            "format": "FORMAT_NONE",
            "dims": [
                -1
            ],
            "is_shape_tensor": false,
            "allow_ragged_batch": false
        },
        {
            "name": "Mean_Humidity",
            "data_type": "TYPE_INT64",
            "format": "FORMAT_NONE",
            "dims": [
                -1
            ],
            "is_shape_tensor": false,
            "allow_ragged_batch": false
        },
        {
            "name": "Min_Humidity",
            "data_type": "TYPE_INT64",
            "format": "FORMAT_NONE",
            "dims": [
                -1
            ],
            "is_shape_tensor": false,
            "allow_ragged_batch": false
        },
        {
            "name": "Max_Wind_SpeedKm_h",
            "data_type": "TYPE_INT64",
            "format": "FORMAT_NONE",
            "dims": [
                -1
            ],
            "is_shape_tensor": false,
            "allow_ragged_batch": false
        },
        {
            "name": "Mean_Wind_SpeedKm_h",
            "data_type": "TYPE_INT64",
            "format": "FORMAT_NONE",
            "dims": [
                -1
            ],
            "is_shape_tensor": false,
            "allow_ragged_batch": false
        },
        {
            "name": "CloudCover",
            "data_type": "TYPE_FP64",
            "format": "FORMAT_NONE",
            "dims": [
                -1
            ],
            "is_shape_tensor": false,
            "allow_ragged_batch": false
        },
        {
            "name": "trend",
            "data_type": "TYPE_INT64",
            "format": "FORMAT_NONE",
            "dims": [
                -1
            ],
            "is_shape_tensor": false,
            "allow_ragged_batch": false
        },
        {
            "name": "trend_DE",
            "data_type": "TYPE_INT64",
            "format": "FORMAT_NONE",
            "dims": [
                -1
            ],
            "is_shape_tensor": false,
            "allow_ragged_batch": false
        },
        {
            "name": "AfterStateHoliday",
            "data_type": "TYPE_FP64",
            "format": "FORMAT_NONE",
            "dims": [
                -1
            ],
            "is_shape_tensor": false,
            "allow_ragged_batch": false
        },
        {
            "name": "BeforeStateHoliday",
            "data_type": "TYPE_FP64",
            "format": "FORMAT_NONE",
            "dims": [
                -1
            ],
            "is_shape_tensor": false,
            "allow_ragged_batch": false
        },
        {
            "name": "Promo",
            "data_type": "TYPE_INT64",
            "format": "FORMAT_NONE",
            "dims": [
                -1
            ],
            "is_shape_tensor": false,
            "allow_ragged_batch": false
        },
        {
            "name": "SchoolHoliday",
            "data_type": "TYPE_INT64",
            "format": "FORMAT_NONE",
            "dims": [
                -1
            ],
            "is_shape_tensor": false,
            "allow_ragged_batch": false
        },
        {
            "name": "Sales",
            "data_type": "TYPE_INT64",
            "format": "FORMAT_NONE",
            "dims": [
                -1
            ],
            "is_shape_tensor": false,
            "allow_ragged_batch": false
        }
    ],
    "output": [
        {
            "name": "Store",
            "data_type": "TYPE_INT64",
            "dims": [
                -1
            ],
            "label_filename": "",
            "is_shape_tensor": false
        },
        {
            "name": "DayOfWeek",
            "data_type": "TYPE_INT64",
            "dims": [
                -1
            ],
            "label_filename": "",
            "is_shape_tensor": false
        },
        {
            "name": "Year",
            "data_type": "TYPE_INT64",
            "dims": [
                -1
            ],
            "label_filename": "",
            "is_shape_tensor": false
        },
        {
            "name": "Month",
            "data_type": "TYPE_INT64",
            "dims": [
                -1
            ],
            "label_filename": "",
            "is_shape_tensor": false
        },
        {
            "name": "Day",
            "data_type": "TYPE_INT64",
            "dims": [
                -1
            ],
            "label_filename": "",
            "is_shape_tensor": false
        },
        {
            "name": "StateHoliday",
            "data_type": "TYPE_INT64",
            "dims": [
                -1
            ],
            "label_filename": "",
            "is_shape_tensor": false
        },
        {
            "name": "CompetitionMonthsOpen",
            "data_type": "TYPE_INT64",
            "dims": [
                -1
            ],
            "label_filename": "",
            "is_shape_tensor": false
        },
        {
            "name": "Promo2Weeks",
            "data_type": "TYPE_INT64",
            "dims": [
                -1
            ],
            "label_filename": "",
            "is_shape_tensor": false
        },
        {
            "name": "StoreType",
            "data_type": "TYPE_INT64",
            "dims": [
                -1
            ],
            "label_filename": "",
            "is_shape_tensor": false
        },
        {
            "name": "Assortment",
            "data_type": "TYPE_INT64",
            "dims": [
                -1
            ],
            "label_filename": "",
            "is_shape_tensor": false
        },
        {
            "name": "PromoInterval",
            "data_type": "TYPE_INT64",
            "dims": [
                -1
            ],
            "label_filename": "",
            "is_shape_tensor": false
        },
        {
            "name": "CompetitionOpenSinceYear",
            "data_type": "TYPE_INT64",
            "dims": [
                -1
            ],
            "label_filename": "",
            "is_shape_tensor": false
        },
        {
            "name": "Promo2SinceYear",
            "data_type": "TYPE_INT64",
            "dims": [
                -1
            ],
            "label_filename": "",
            "is_shape_tensor": false
        },
        {
            "name": "State",
            "data_type": "TYPE_INT64",
            "dims": [
                -1
            ],
            "label_filename": "",
            "is_shape_tensor": false
        },
        {
            "name": "Week",
            "data_type": "TYPE_INT64",
            "dims": [
                -1
            ],
            "label_filename": "",
            "is_shape_tensor": false
        },
        {
            "name": "Events",
            "data_type": "TYPE_INT64",
            "dims": [
                -1
            ],
            "label_filename": "",
            "is_shape_tensor": false
        },
        {
            "name": "Promo_fw",
            "data_type": "TYPE_INT64",
            "dims": [
                -1
            ],
            "label_filename": "",
            "is_shape_tensor": false
        },
        {
            "name": "Promo_bw",
            "data_type": "TYPE_INT64",
            "dims": [
                -1
            ],
            "label_filename": "",
            "is_shape_tensor": false
        },
        {
            "name": "StateHoliday_fw",
            "data_type": "TYPE_INT64",
            "dims": [
                -1
            ],
            "label_filename": "",
            "is_shape_tensor": false
        },
        {
            "name": "StateHoliday_bw",
            "data_type": "TYPE_INT64",
            "dims": [
                -1
            ],
            "label_filename": "",
            "is_shape_tensor": false
        },
        {
            "name": "SchoolHoliday_fw",
            "data_type": "TYPE_INT64",
            "dims": [
                -1
            ],
            "label_filename": "",
            "is_shape_tensor": false
        },
        {
            "name": "SchoolHoliday_bw",
            "data_type": "TYPE_INT64",
            "dims": [
                -1
            ],
            "label_filename": "",
            "is_shape_tensor": false
        },
        {
            "name": "CompetitionDistance",
            "data_type": "TYPE_FP32",
            "dims": [
                -1
            ],
            "label_filename": "",
            "is_shape_tensor": false
        },
        {
            "name": "Max_TemperatureC",
            "data_type": "TYPE_FP32",
            "dims": [
                -1
            ],
            "label_filename": "",
            "is_shape_tensor": false
        },
        {
            "name": "Mean_TemperatureC",
            "data_type": "TYPE_FP32",
            "dims": [
                -1
            ],
            "label_filename": "",
            "is_shape_tensor": false
        },
        {
            "name": "Min_TemperatureC",
            "data_type": "TYPE_FP32",
            "dims": [
                -1
            ],
            "label_filename": "",
            "is_shape_tensor": false
        },
        {
            "name": "Max_Humidity",
            "data_type": "TYPE_FP32",
            "dims": [
                -1
            ],
            "label_filename": "",
            "is_shape_tensor": false
        },
        {
            "name": "Mean_Humidity",
            "data_type": "TYPE_FP32",
            "dims": [
                -1
            ],
            "label_filename": "",
            "is_shape_tensor": false
        },
        {
            "name": "Min_Humidity",
            "data_type": "TYPE_FP32",
            "dims": [
                -1
            ],
            "label_filename": "",
            "is_shape_tensor": false
        },
        {
            "name": "Max_Wind_SpeedKm_h",
            "data_type": "TYPE_FP32",
            "dims": [
                -1
            ],
            "label_filename": "",
            "is_shape_tensor": false
        },
        {
            "name": "Mean_Wind_SpeedKm_h",
            "data_type": "TYPE_FP32",
            "dims": [
                -1
            ],
            "label_filename": "",
            "is_shape_tensor": false
        },
        {
            "name": "CloudCover",
            "data_type": "TYPE_FP32",
            "dims": [
                -1
            ],
            "label_filename": "",
            "is_shape_tensor": false
        },
        {
            "name": "trend",
            "data_type": "TYPE_FP32",
            "dims": [
                -1
            ],
            "label_filename": "",
            "is_shape_tensor": false
        },
        {
            "name": "trend_DE",
            "data_type": "TYPE_FP32",
            "dims": [
                -1
            ],
            "label_filename": "",
            "is_shape_tensor": false
        },
        {
            "name": "AfterStateHoliday",
            "data_type": "TYPE_FP32",
            "dims": [
                -1
            ],
            "label_filename": "",
            "is_shape_tensor": false
        },
        {
            "name": "BeforeStateHoliday",
            "data_type": "TYPE_FP32",
            "dims": [
                -1
            ],
            "label_filename": "",
            "is_shape_tensor": false
        },
        {
            "name": "Promo",
            "data_type": "TYPE_FP32",
            "dims": [
                -1
            ],
            "label_filename": "",
            "is_shape_tensor": false
        },
        {
            "name": "SchoolHoliday",
            "data_type": "TYPE_FP32",
            "dims": [
                -1
            ],
            "label_filename": "",
            "is_shape_tensor": false
        },
        {
            "name": "Sales",
            "data_type": "TYPE_FP32",
            "dims": [
                -1
            ],
            "label_filename": "",
            "is_shape_tensor": false
        }
    ],
    "batch_input": [],
    "batch_output": [],
    "optimization": {
        "priority": "PRIORITY_DEFAULT",
        "input_pinned_memory": {
            "enable": true
        },
        "output_pinned_memory": {
            "enable": true
        }
    },
    "instance_group": [
        {
            "name": "rossmann_categorify",
            "kind": "KIND_GPU",
            "count": 1,
            "gpus": [
                0
            ],
            "profile": []
        }
    ],
    "default_model_filename": "",
    "cc_model_filenames": {},
    "metric_tags": {},
    "parameters": {},
    "model_warmup": []
}
I0202 19:39:20.501040 1083 backend.cc:510] TRITONBACKEND_ModelInstanceInitialize: rossmann_categorify (GPU device 0)
I0202 19:39:20.528417 1083 nvtabular.h:86] Python interpreter is initialized
terminate called after throwing an instance of 'pybind11::error_already_set'
  what():  ModuleNotFoundError: No module named 'nvt'
Aborted (core dumped)
```
</p>
</details>

Reply:
I'm also thinking we should probably merge this as is - given that this is just to get working with the python backend (which works) - and then work on integrating with the C++ backend in a future PR. 

Minimally we will also need to change the 'backend' in the config.pbtxt file from "python" to "nvtabular" or some such.